### ISSUE TITLE
Three separate calls required to upload a patient attachment

### ISSUE THREAD

**Issue by hagen** [redguava/cliniko-api#196]
Hi guys, I know this isn't in master of the docs yet, but I just wanted to walk through the logic behind this.

To upload an attachment to Cliniko, each attachment requires three separate HTTP requests from an integrator. Actually more, including polling to check when the file is processed. Could you not do the S3 upload as part of the patient_attachments POST request?
e.g. we POST
```json
{
    "content": "<Buffer>",
    "file_name": "intake form.pdf",
    "content_type": "application/pdf",
    "patient_id": 12345678
}
```
and Cliniko pushes to S3, then commences processing, returning to the caller a payload indicating the file is now processing?

**Comment by bpinto** [comment]
Hey Hagen, we are going to review this end point before it's final, but I don't think we would be able to do as you described.

We are currently using these different HTTP requests so that your upload goes directly to S3 instead of blocking one of our servers for the upload. The way the integration has been built also allow us to use another provider in the future if we decide to.

It would also allow Cliniko to accept uploads from another server directly (by specifying a URL) in the future - we don't intend to support this for now.

**Comment by hagen** [comment]
Thanks Bruno,
that makes sense. I assumed it would have been to avoid passing the document via. your internals - lots of data for no real purpose.
I mean it works and all, I have no problems with that, but you could maybe have an elastic cluster of S3 uploaders, and potentially re-route all API requests with `patient_attachments` in them to that cluster.

I guess I'm just concerned that - as far as your endpoints go - this one seems unnecessarily complicated for integrators to work with.

I take your point that because you're pulling the file from a remote location (in this case, a different bucket) you can ingest files from any public location, but that doesn't seem like a good justification for implementing the same in the API. We have to make the call to upload a document anyway - why don't we just POST it to Cliniko initially, instead of you then fetching it from elsewhere?

But as I'm typing this, I'm thinking maybe it's done this way so there's the opportunity to pull from services like Dropbox... that would make sense. But still seems odd that we have to mimic that behaviour to get an attachment into Cliniko via. the API.

Anyway, I'm not overly fussed, but seems this could work a little more succinctly.

Thanks for all the hard work getting to this stage! Looking forward to upload functionality, however it comes :)

**Comment by bpinto** [comment]
Having a complex end point is not a good thing. We've always done a lot of changes to our app structure before exposing a new API end point to guarantee the end point would be easy to use. I think we can say that once you know how to use one end point you can use most of the other end points without much problems.

At this moment, if we decide to expose the attachment in a different way (completely hiding its S3 backend) we would have to postpone this end point release for a long time, an approach we've had with other end points in the past.

If we release this end point as is, we could still create an easier end point in the future, when we have man power to do this, unfortunately that also means we need to support this older end point. :P

We'll have more news soon.

**Comment by JohnColvin** [comment]
We've discussed this more and we're going to continue with the current implementation. People want this badly and we're looking at a release this week as long as no other issues pop up.

When we first started reviewing the file upload process, I was of the same mind that 3 requests and handling s3 fields was too complicated. However, once I started to work with it, I think it makes sense. For us, the critical piece here is that Cliniko servers are not handling the upload. When you make that the immovable requirement of this, it becomes much harder to make a single endpoint to do this.

The 3 request implementation is chatty, but it is self describing and leaves the door open to change storage providers or accept files from other sources. This is the same process that we use in JS in the Cliniko UI and it's not all that much code.

I think what we have is the current right balance of preventing server load and having attachments in the API.
